flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -0,0 +1,287 @@
+			List<SchedulingResultPartition> finishedPartitions = null;
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking()) {
+					SchedulingIntermediateDataSet intermediateDataSet = intermediateDataSets.get(srp.getResultId());
+					if (intermediateDataSet == null) {
+						throw new IllegalStateException("can not find scheduling intermediate dataset for "
+							+ srp.getResultId());
+					}
+					if (intermediateDataSet.markPartitionFinished(srp.getId())) {
+						if (finishedPartitions == null) {
+							finishedPartitions = new ArrayList<>();
+						}
+						finishedPartitions.addAll(intermediateDataSet.getSchedulingResultPartition());
+					}
+				}
+			}
+
+			if (finishedPartitions != null) {
+				List<SchedulingExecutionVertex> toScheduleConsumers = new ArrayList<>();
+				for (SchedulingResultPartition finishedPartition : finishedPartitions) {
+					toScheduleConsumers.addAll(finishedPartition.getConsumers());
+				}
+				scheduleBlockingConsumers(toScheduleConsumers);
+			}
+		}
+	}
+
+	@Override
+	public void onPartitionConsumable(ExecutionVertexID executionVertexId, ResultPartitionID resultPartitionId) {
+		final SchedulingExecutionVertex producerVertex = schedulingTopology.getVertex(executionVertexId)
+			.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
+		final SchedulingResultPartition resultPartition = schedulingTopology
+			.getResultPartition(resultPartitionId.getPartitionId())
+			.orElseThrow(() -> new IllegalStateException("can not find scheduling result partition for "
+				+ resultPartitionId));
+
+		if (!producerVertex.getProducedResultPartitions().contains(resultPartition)) {
+			throw new IllegalStateException("partition " + resultPartitionId
+				+ " is not the produced partition of " + executionVertexId);
+		}
+
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>();
+		for (SchedulingExecutionVertex schedulingVertex : resultPartition.getConsumers()) {
+			if (ExecutionState.CREATED.equals(schedulingVertex.getState()) &&
+				resultPartition.getPartitionType().isPipelined()) {
+				executionVertexDeploymentOptions.add(
+					new ExecutionVertexDeploymentOption(schedulingVertex.getId(),
+						deploymentOptions.get(schedulingVertex.getId())));
+			}
+		}
+		schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+	}
+
+	private boolean partitionConsumable(SchedulingResultPartition partition) {
+		SchedulingResultPartition.ResultPartitionState state = partition.getState();
+		if (partition.getPartitionType().isPipelined()) {
+			return PRODUCING.equals(state) || DONE.equals(state);
+		} else {
+			SchedulingIntermediateDataSet intermediateDataSet = intermediateDataSets.get(partition.getResultId());
+			if (intermediateDataSet == null) {
+				throw new IllegalStateException("can not find scheduling intermediate dataset for "
```

Comment:
This check is duplicated. See my other comment in `LazyFromSourcesSchedulingStrategyTest` for a possible solution.

Reply:
OK, I will wrap it in `SchedulingIntermediateDataSetManager`.